pkg/pool-stable/contracts/meta/StableOracleMath.sol

```diff
@@ -0,0 +1,118 @@
+        //                                                                                                           //
+        // n = 2                                                                                                     //
+        // a = amp param * n                                                                                         //
+        // b = D + a.(S - D)                                                                                         //
+        // D = invariant                                                                                             //
+        // S = sum of balances but x,y = 0                                                                           //
+        **************************************************************************************************************/
+
+        uint256 invariant = _calculateInvariant(amplificationParameter, _balances(balanceX, balanceY), true);
+
+        uint256 a = (amplificationParameter * 2) / _AMP_PRECISION;
+        uint256 b = Math.mul(invariant, a).sub(invariant);
+
+        uint256 axy2 = Math.mul(a * 2, balanceX).mulUp(balanceY);
+
+        // dx = a.x.y.2 + a.y^2 - b.y
+        uint256 derivativeX = axy2.add(Math.mul(a, balanceY).mulUp(balanceY)).sub(b.mulUp(balanceY));
+
+        // dy = a.x.y.2 + a.x^2 - b.x
+        uint256 derivativeY = axy2.add(Math.mul(a, balanceX).mulDown(balanceX)).sub(b.mulDown(balanceX));
+
+        // The rounding direction is irrelevant as we're about to introduce a much larger error when converting to log
+        // space. We use `divUp` as it prevents the result from being zero, which would make the logarithm revert. A
+        // result of zero is therefore only possible with zero balances, which are prevented via other means.
+        return derivativeX.divUp(derivativeY);
+    }
+
+    /**
+     * @dev Calculates the price of BPT in token. A `logBptTotalSupply` should be the result of calling
+     * `LogCompression.toLowResLog` with the current BPT supply.
+     *
+     * The return value is a 4 decimal fixed-point number: use `LogCompression.fromLowResLog`
+     * to recover the original value.
+     */
+    function _calcLogBptPrice(
+        uint256 spotPrice,
+        uint256 balanceX,
+        uint256 balanceY,
+        int256 logBptTotalSupply
+    ) internal pure returns (int256) {
+        /**************************************************************************************************************
+        //                                                                                                           //
+        //              balance X + (spot price Y/X * balance Y)                                                     //
+        // BPT price = ------------------------------------------                                                    //
+        //                          total supply                                                                     //
```

Comment:
I don't follow this formula, where does the derivation come from?

Reply:
It's simply transforming the balance of token Y into X using the spot price and then dividing the supply, that basically tells you how much your supply means in X

Reply:
Ah I see, the formula the price confused me. So the spot price is the price of Y in terms of X, meaning it has units [X/Y], and when multiplied by a Y amount you get an equivalent X amount. Thanks!